cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -0,0 +1,361 @@
+	} else if opts.ClusterName != "" && cfg.Global.ClusterName == "" {
+		cfg.Global.ClusterName = opts.ClusterName
+	}
+
+	manager := packetManagerRest{
+		clusterName:       cfg.Global.ClusterName,
+		projectID:         cfg.Global.ProjectID,
+		apiServerEndpoint: cfg.Global.APIServerEndpoint,
+		facility:          cfg.Global.Facility,
+		plan:              cfg.Global.Plan,
+		os:                cfg.Global.OS,
+		billing:           cfg.Global.Billing,
+		cloudinit:         cfg.Global.CloudInit,
+		reservation:       cfg.Global.Reservation,
+	}
+	return &manager, nil
+}
+
+func (mgr *packetManagerRest) listPacketDevices() *Devices {
+	var jsonStr = []byte(``)
+	packetAuthToken := os.Getenv("PACKET_AUTH_TOKEN")
+	url := "https://api.packet.net/projects/" + mgr.projectID + "/devices"
+	req, err := http.NewRequest("GET", url, bytes.NewBuffer(jsonStr))
+	req.Header.Set("X-Auth-Token", packetAuthToken)
+	req.Header.Set("Content-Type", "application/json")
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+		// klog.Fatalf("Error listing nodes: %v", err)
+	}
+	defer resp.Body.Close()
+
+	fmt.Println("response Status:", resp.Status)
+	body, _ := ioutil.ReadAll(resp.Body)
+	var devices Devices
+	json.Unmarshal([]byte(body), &devices)
+	return &devices
+}
+
+// nodeGroupSize gets the current size of the nodegroup as reported by packet tags.
+func (mgr *packetManagerRest) nodeGroupSize(nodegroup string) (int, error) {
+	devices := mgr.listPacketDevices()
+	// Get the count of devices tagged as nodegroup members
+	count := 0
+	for _, d := range devices.Devices {
+		if Contains(d.Tags, "k8s-cluster-"+mgr.clusterName) && Contains(d.Tags, "k8s-nodepool-"+nodegroup) {
+			count++
+		}
+	}
+	fmt.Println(len(devices.Devices), count)
+	return count, nil
+}
+
+func randString8() string {
+	n := 8
+	rand.Seed(time.Now().UnixNano())
+	letterRunes := []rune("acdefghijklmnopqrstuvwxyz")
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(b)
+}
+
+// createNodes provisions new nodes on packet and bootstraps them in the cluster.
+func (mgr *packetManagerRest) createNodes(nodegroup string, nodes int) error {
+	klog.Infof("Updating node count to %d for nodegroup %s", nodes, nodegroup)
+	cloudinit, err := base64.StdEncoding.DecodeString(mgr.cloudinit)
+	if err != nil {
+		log.Fatal(err)
+		return fmt.Errorf("Could not decode cloudinit script: %v", err)
+	}
+
+	udvars := CloudInitTemplateData{
+		BootstrapTokenID:     os.Getenv("BOOTSTRAP_TOKEN_ID"),
+		BootstrapTokenSecret: os.Getenv("BOOTSTRAP_TOKEN_SECRET"),
+		APIServerEndpoint:    mgr.apiServerEndpoint,
+	}
+	ud := renderTemplate(string(cloudinit), udvars)
+	hn := "k8s-" + mgr.clusterName + "-" + nodegroup + "-" + randString8()
+
+	reservation := ""
+	if mgr.reservation == "require" || mgr.reservation == "prefer" {
+		reservation = "next-available"
+	}
+
+	cr := DeviceCreateRequest{
```

Comment:
Hi Rodrigo. Your suggestion makes sense to me. However, I think it was in our call the other day when somebody (perhaps @deitch?) mentioned that it's already possible to support multiple nodepools by running multiple instances of cluster-autoscaler. This makes sense to me as well and requires no extra dev ;)

As a developer I'd rather keep complexity low. On the other hand, as a cluster-autoscaler user, I'd much rather configure it once for multiple nodepools. So, I tend to agree that it's probably worth the extra dev effort and complexity. That said, it's not a trivial change and it doesn't feel like a particularly urgent requirement, so I'd probably schedule this enhancement for the next iteration.

Reply:
It was not I on the call, @d-mo , but you can do that. It feels a little "weak" compared to what you do on hyperscalers. E.g. when I run things in AWS (yes, I sometimes do stuff not on Packet, but only when someone insists! :-) ), I set the ASG discovery, it can pick up tags on multiple ASG, and manage each as a separate nodepool.

We _should_ do the same here, i.e. have it have multiple nodepools. 

I do agree that it is a v2 (or v0.2, etc.) request. Let's get it in working with just one nodepool, and immediately open an issue on it requesting multiple nodepools.

Reply:
Thanks a lot for the answer, to both! :)

Deploying the autoscaler for each pool (we have installations with 10 or more pools :-/) is not quite a good experience to use it in production, and we need to autoscale almost all of them.

I understand the concern from a dev POV, and it makes sense to start small. I'll let you decide when to support more than one pool :)

Just curious, do you plan to do the next iteration soon? :)